pkg/ingester/client/timeseries.go

```diff
@@ -14,6 +15,27 @@ var (
 	expectedTimeseries       = 100
 	expectedLabels           = 20
 	expectedSamplesPerSeries = 10
+
+	// TSPool is the pool to return the timeseries to.
+	TSPool = sync.Pool{
+		New: func() interface{} {
+			return make([]PreallocTimeseries, 0, expectedTimeseries)
+		},
+	}
+
+	// LabelAdapterPool is the pool to return the labels to.
+	LabelAdapterPool = sync.Pool{
+		New: func() interface{} {
+			return make([]LabelAdapter, 0, expectedLabels)
+		},
+	}
+
+	// SamplePool is the pool to return the samples to.
+	SamplePool = sync.Pool{
+		New: func() interface{} {
+			return make([]Sample, 0, expectedSamplesPerSeries)
+		},
+	}
```

Comment:
Do we need 3 pools here?  Or can we get away with a single pool of TimeSeries, as they hold references to the Labels and Samples?

Reply:
So, I think we'll still need 3 pools because, in the unmarshalling code we have:

https://github.com/cortexproject/cortex/blob/master/pkg/ingester/client/cortex.pb.go#L4515-L4518

This'll create a new `Timeseries` and append it. We need the preallocs happening on Unmarshal. Not sure how I can do that with a single pool.

Reply:
This is just a thought (you're probably right): in Timeseries.Unmarshall, if the slices are nill, we allocate new non-zero length ones.

These newly allocated timeseries will be added to slices that eventually end up in the pool (via the ReuseTS function).

I might test this out to see if it has the same effect.   Given remote write sends a fixes number of series, and we prealloc that number, I suspect the line you reference is never called.